The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6363504688/job/17279165291) [(plain)](https://github.com/rust-lang/rust/commit/556e0bf75712cf6095dac25d2bd6d6be06a4dba5/checks/17279165291/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=ouz-a
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_938b69fe-050d-4c33-8fc0-1490bf926818
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=smir_spans
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_938b69fe-050d-4c33-8fc0-1490bf926818
GITHUB_REF=refs/pull/116288/merge
GITHUB_REF_NAME=116288/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=a4e73263c583004bd062536497f5d953261827cb
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_938b69fe-050d-4c33-8fc0-1490bf926818
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_938b69fe-050d-4c33-8fc0-1490bf926818
GITHUB_TRIGGERING_ACTOR=ouz-a
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116288/merge
GITHUB_WORKFLOW_SHA=a4e73263c583004bd062536497f5d953261827cb
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
   |
LL |         stable_mir::mir::Terminator::Call { .. } => {}
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: if there were a trait named `Example` with associated type `Call` implemented for `stable_mir::mir::Terminator`, you could use the fully-qualified path
   |
LL |         <stable_mir::mir::Terminator as Example>::Call { .. } => {}

error[E0599]: no method named `kind` found for struct `stable_mir::mir::LocalDecl` in the current scope
##[error]  --> /checkout/tests/ui-fulldeps/stable-mir/crate-info.rs:73:24
   |
   |
LL |         body.locals[0].kind(),
   |                        ^^^^ method not found in `LocalDecl`
help: one of the expressions' fields has a method of the same name
   |
   |
LL |         body.locals[0].ty.kind(),

error[E0599]: no method named `kind` found for struct `stable_mir::mir::LocalDecl` in the current scope
##[error]  --> /checkout/tests/ui-fulldeps/stable-mir/crate-info.rs:77:24
   |
   |
LL |         body.locals[1].kind(),
   |                        ^^^^ method not found in `LocalDecl`
help: one of the expressions' fields has a method of the same name
   |
   |
LL |         body.locals[1].ty.kind(),

error[E0599]: no method named `kind` found for struct `stable_mir::mir::LocalDecl` in the current scope
##[error]  --> /checkout/tests/ui-fulldeps/stable-mir/crate-info.rs:81:24
   |
   |
LL |         body.locals[2].kind(),
   |                        ^^^^ method not found in `LocalDecl`
help: one of the expressions' fields has a method of the same name
   |
   |
LL |         body.locals[2].ty.kind(),

error[E0599]: no method named `kind` found for struct `stable_mir::mir::LocalDecl` in the current scope
##[error]  --> /checkout/tests/ui-fulldeps/stable-mir/crate-info.rs:85:24
   |
   |
LL |         body.locals[3].kind(),
   |                        ^^^^ method not found in `LocalDecl`
help: one of the expressions' fields has a method of the same name
   |
   |
LL |         body.locals[3].ty.kind(),

error[E0599]: no method named `kind` found for struct `stable_mir::mir::LocalDecl` in the current scope
##[error]  --> /checkout/tests/ui-fulldeps/stable-mir/crate-info.rs:89:24
   |
   |
LL |         body.locals[4].kind(),
   |                        ^^^^ method not found in `LocalDecl`
help: one of the expressions' fields has a method of the same name
   |
   |
LL |         body.locals[4].ty.kind(),

error[E0599]: no method named `kind` found for struct `stable_mir::mir::LocalDecl` in the current scope
##[error]  --> /checkout/tests/ui-fulldeps/stable-mir/crate-info.rs:93:24
   |
   |
LL |         body.locals[5].kind(),
   |                        ^^^^ method not found in `LocalDecl`
help: one of the expressions' fields has a method of the same name
   |
   |
LL |         body.locals[5].ty.kind(),

error[E0223]: ambiguous associated type
##[error]  --> /checkout/tests/ui-fulldeps/stable-mir/crate-info.rs:104:9
   |
   |
LL |         stable_mir::mir::Terminator::Drop { .. } => {}
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: if there were a trait named `Example` with associated type `Drop` implemented for `stable_mir::mir::Terminator`, you could use the fully-qualified path
   |
LL |         <stable_mir::mir::Terminator as Example>::Drop { .. } => {}

error[E0223]: ambiguous associated type
##[error]  --> /checkout/tests/ui-fulldeps/stable-mir/crate-info.rs:113:9
   |
   |
LL |         stable_mir::mir::Terminator::Assert { .. } => {}
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: if there were a trait named `Example` with associated type `Assert` implemented for `stable_mir::mir::Terminator`, you could use the fully-qualified path
   |
LL |         <stable_mir::mir::Terminator as Example>::Assert { .. } => {}

error[E0223]: ambiguous associated type
##[error]  --> /checkout/tests/ui-fulldeps/stable-mir/crate-info.rs:120:13
   |
   |
LL |             stable_mir::mir::Terminator::Call { func, .. } => match func {
   |
   |
help: if there were a trait named `Example` with associated type `Call` implemented for `stable_mir::mir::Terminator`, you could use the fully-qualified path
   |
LL |             <stable_mir::mir::Terminator as Example>::Call { func, .. } => match func {

error[E0599]: no associated item named `Return` found for struct `stable_mir::mir::Terminator` in the current scope
##[error]  --> /checkout/tests/ui-fulldeps/stable-mir/crate-info.rs:152:42
   |
---
##[error]  --> /checkout/tests/ui-fulldeps/stable-mir/crate-info.rs:131:38
   |
LL |   ...                   match func.locals[1]
   |  _____________________________-
LL | | ...                       .fold(&mut args)
   | |                           -^^^^ `stable_mir::mir::LocalDecl` is not an iterator
   | 
  --> /rustc/FAKE_PREFIX/compiler/stable_mir/src/mir/body.rs:12:1
   |
   |
   = note: doesn't satisfy `stable_mir::mir::LocalDecl: Iterator`
   = note: the following trait bounds were not satisfied:
   = note: the following trait bounds were not satisfied:
           `stable_mir::mir::LocalDecl: Iterator`
           which is required by `&mut stable_mir::mir::LocalDecl: Iterator`
warning: unused import: `stable_mir::fold::Foldable`
##[warning]  --> /checkout/tests/ui-fulldeps/stable-mir/crate-info.rs:22:5
   |
LL | use stable_mir::fold::Foldable;
```

</details>

--- a/compiler/rustc_smir/src/rustc_smir/mod.rs
+++ b/compiler/rustc_smir/src/rustc_smir/mod.rs
@@ -15,7 +15,7 @@ use rustc_middle::mir::interpret::{alloc_range, AllocId};
 use rustc_middle::ty::{self, Ty, TyCtxt, Variance};
 use rustc_span::def_id::{CrateNum, DefId, LOCAL_CRATE};
 use rustc_target::abi::FieldIdx;
-use stable_mir::mir::{CopyNonOverlapping, UserTypeProjection, VariantIdx};
+use stable_mir::mir::{CopyNonOverlapping, Statement, UserTypeProjection, VariantIdx};
 use stable_mir::ty::{FloatTy, GenericParamDef, IntTy, Movability, RigidTy, Span, TyKind, UintTy};
 use stable_mir::{self, opaque, Context};
 use tracing::debug;
@@ -231,40 +231,71 @@ impl<'tcx> Stable<'tcx> for mir::Statement<'tcx> {
     type T = stable_mir::mir::Statement;
     fn stable(&self, tables: &mut Tables<'tcx>) -> Self::T {
         use rustc_middle::mir::StatementKind::*;
+        let span = self.source_info.span.stable(tables);
         match &self.kind {
-            Assign(assign) => {
-                stable_mir::mir::Statement::Assign(assign.0.stable(tables), assign.1.stable(tables))
-            }
-            FakeRead(fake_read_place) => stable_mir::mir::Statement::FakeRead(
-                fake_read_place.0.stable(tables),
-                fake_read_place.1.stable(tables),
-            ),
-            SetDiscriminant { place: plc, variant_index: idx } => {
-                stable_mir::mir::Statement::SetDiscriminant {
+            Assign(assign) => Statement {
+                kind: stable_mir::mir::StatementKind::Assign(
+                    assign.0.stable(tables),
+                    assign.1.stable(tables),
+                ),
+                span,
+            },
+            FakeRead(fake_read_place) => Statement {
+                kind: stable_mir::mir::StatementKind::FakeRead(
+                    fake_read_place.0.stable(tables),
+                    fake_read_place.1.stable(tables),
+                ),
+                span,
+            },
+            SetDiscriminant { place: plc, variant_index: idx } => Statement {
+                kind: stable_mir::mir::StatementKind::SetDiscriminant {
                     place: plc.as_ref().stable(tables),
                     variant_index: idx.stable(tables),
-                }
-            }
-            Deinit(place) => stable_mir::mir::Statement::Deinit(place.stable(tables)),
-            StorageLive(place) => stable_mir::mir::Statement::StorageLive(place.stable(tables)),
-            StorageDead(place) => stable_mir::mir::Statement::StorageDead(place.stable(tables)),
-            Retag(retag, place) => {
-                stable_mir::mir::Statement::Retag(retag.stable(tables), place.stable(tables))
-            }
-            PlaceMention(place) => stable_mir::mir::Statement::PlaceMention(place.stable(tables)),
-            AscribeUserType(place_projection, variance) => {
-                stable_mir::mir::Statement::AscribeUserType {
+                },
+                span,
+            },
+            Deinit(place) => Statement {
+                kind: stable_mir::mir::StatementKind::Deinit(place.stable(tables)),
+                span,
+            },
+            StorageLive(place) => Statement {
+                kind: stable_mir::mir::StatementKind::StorageLive(place.stable(tables)),
+                span,
+            },
+            StorageDead(place) => Statement {
+                kind: stable_mir::mir::StatementKind::StorageDead(place.stable(tables)),
+                span,
+            },
+            Retag(retag, place) => Statement {
+                kind: stable_mir::mir::StatementKind::Retag(
+                    retag.stable(tables),
+                    place.stable(tables),
+                ),
+                span,
+            },
+            PlaceMention(place) => Statement {
+                kind: stable_mir::mir::StatementKind::PlaceMention(place.stable(tables)),
+                span,
+            },
+            AscribeUserType(place_projection, variance) => Statement {
+                kind: stable_mir::mir::StatementKind::AscribeUserType {
                     place: place_projection.as_ref().0.stable(tables),
                     projections: place_projection.as_ref().1.stable(tables),
                     variance: variance.stable(tables),
-                }
+                },
+                span,
+            },
+            Coverage(coverage) => {
+                Statement { kind: stable_mir::mir::StatementKind::Coverage(opaque(coverage)), span }
             }
-            Coverage(coverage) => stable_mir::mir::Statement::Coverage(opaque(coverage)),
-            Intrinsic(intrinstic) => {
-                stable_mir::mir::Statement::Intrinsic(intrinstic.stable(tables))
+            Intrinsic(intrinstic) => Statement {
+                kind: stable_mir::mir::StatementKind::Intrinsic(intrinstic.stable(tables)),
+                span,
+            },
+            ConstEvalCounter => {
+                Statement { kind: stable_mir::mir::StatementKind::ConstEvalCounter, span }
             }
-            ConstEvalCounter => stable_mir::mir::Statement::ConstEvalCounter,
-            Nop => stable_mir::mir::Statement::Nop,
+            Nop => Statement { kind: stable_mir::mir::StatementKind::Nop, span },
         }
     }
 }
@@ -816,11 +847,11 @@ impl<'tcx> Stable<'tcx> for mir::Terminator<'tcx> {
         use rustc_middle::mir::TerminatorKind::*;
         use stable_mir::mir::Terminator;
         use stable_mir::mir::TerminatorKind;
+        let span = self.source_info.span.stable(tables);
         match &self.kind {
-            Goto { target } => Terminator {
-                kind: TerminatorKind::Goto { target: target.as_usize() },
-                span: self.source_info.span.stable(tables),
-            },
+            Goto { target } => {
+                Terminator { kind: TerminatorKind::Goto { target: target.as_usize() }, span }
+            }
             SwitchInt { discr, targets } => Terminator {
                 kind: TerminatorKind::SwitchInt {
                     discr: discr.stable(tables),
@@ -833,31 +864,19 @@ impl<'tcx> Stable<'tcx> for mir::Terminator<'tcx> {
                         .collect(),
                     otherwise: targets.otherwise().as_usize(),
                 },
-                span: self.source_info.span.stable(tables),
+                span,
             },
-            UnwindResume => Terminator {
-                kind: TerminatorKind::Resume,
-                span: self.source_info.span.stable(tables),
-            },
-            UnwindTerminate(_) => Terminator {
-                kind: TerminatorKind::Abort,
-                span: self.source_info.span.stable(tables),
-            },
-            Return => Terminator {
-                kind: TerminatorKind::Return,
-                span: self.source_info.span.stable(tables),
-            },
-            Unreachable => Terminator {
-                kind: TerminatorKind::Unreachable,
-                span: self.source_info.span.stable(tables),
-            },
+            UnwindResume => Terminator { kind: TerminatorKind::Resume, span },
+            UnwindTerminate(_) => Terminator { kind: TerminatorKind::Abort, span },
+            Return => Terminator { kind: TerminatorKind::Return, span },
+            Unreachable => Terminator { kind: TerminatorKind::Unreachable, span },
             Drop { place, target, unwind, replace: _ } => Terminator {
                 kind: TerminatorKind::Drop {
                     place: place.stable(tables),
                     target: target.as_usize(),
                     unwind: unwind.stable(tables),
                 },
-                span: self.source_info.span.stable(tables),
+                span,
             },
             Call { func, args, destination, target, unwind, call_source: _, fn_span: _ } => {
                 Terminator {
@@ -868,7 +887,7 @@ impl<'tcx> Stable<'tcx> for mir::Terminator<'tcx> {
                         target: target.map(|t| t.as_usize()),
                         unwind: unwind.stable(tables),
                     },
-                    span: self.source_info.span.stable(tables),
+                    span,
                 }
             }
             Assert { cond, expected, msg, target, unwind } => Terminator {
@@ -879,7 +898,7 @@ impl<'tcx> Stable<'tcx> for mir::Terminator<'tcx> {
                     target: target.as_usize(),
                     unwind: unwind.stable(tables),
                 },
-                span: self.source_info.span.stable(tables),
+                span,
             },
             InlineAsm { template, operands, options, line_spans, destination, unwind } => {
                 Terminator {
@@ -891,7 +910,7 @@ impl<'tcx> Stable<'tcx> for mir::Terminator<'tcx> {
                         destination: destination.map(|d| d.as_usize()),
                         unwind: unwind.stable(tables),
                     },
-                    span: self.source_info.span.stable(tables),
+                    span,
                 }
             }
             Yield { .. } | GeneratorDrop | FalseEdge { .. } | FalseUnwind { .. } => unreachable!(),

--- a/compiler/rustc_smir/src/rustc_smir/mod.rs
+++ b/compiler/rustc_smir/src/rustc_smir/mod.rs
@@ -230,72 +230,64 @@ pub(crate) trait Stable<'tcx> {
 impl<'tcx> Stable<'tcx> for mir::Statement<'tcx> {
     type T = stable_mir::mir::Statement;
     fn stable(&self, tables: &mut Tables<'tcx>) -> Self::T {
-        use rustc_middle::mir::StatementKind::*;
-        let span = self.source_info.span.stable(tables);
-        match &self.kind {
-            Assign(assign) => Statement {
-                kind: stable_mir::mir::StatementKind::Assign(
-                    assign.0.stable(tables),
-                    assign.1.stable(tables),
-                ),
-                span,
-            },
-            FakeRead(fake_read_place) => Statement {
-                kind: stable_mir::mir::StatementKind::FakeRead(
+        Statement { kind: self.kind.stable(tables), span: self.source_info.span.stable(tables) }
+    }
+}
+
+impl<'tcx> Stable<'tcx> for mir::StatementKind<'tcx> {
+    type T = stable_mir::mir::StatementKind;
+    fn stable(&self, tables: &mut Tables<'tcx>) -> Self::T {
+        match self {
+            mir::StatementKind::Assign(assign) => stable_mir::mir::StatementKind::Assign(
+                assign.0.stable(tables),
+                assign.1.stable(tables),
+            ),
+            mir::StatementKind::FakeRead(fake_read_place) => {
+                stable_mir::mir::StatementKind::FakeRead(
                     fake_read_place.0.stable(tables),
                     fake_read_place.1.stable(tables),
-                ),
-                span,
-            },
-            SetDiscriminant { place: plc, variant_index: idx } => Statement {
-                kind: stable_mir::mir::StatementKind::SetDiscriminant {
-                    place: plc.as_ref().stable(tables),
-                    variant_index: idx.stable(tables),
-                },
-                span,
-            },
-            Deinit(place) => Statement {
-                kind: stable_mir::mir::StatementKind::Deinit(place.stable(tables)),
-                span,
-            },
-            StorageLive(place) => Statement {
-                kind: stable_mir::mir::StatementKind::StorageLive(place.stable(tables)),
-                span,
-            },
-            StorageDead(place) => Statement {
-                kind: stable_mir::mir::StatementKind::StorageDead(place.stable(tables)),
-                span,
-            },
-            Retag(retag, place) => Statement {
-                kind: stable_mir::mir::StatementKind::Retag(
-                    retag.stable(tables),
-                    place.stable(tables),
-                ),
-                span,
-            },
-            PlaceMention(place) => Statement {
-                kind: stable_mir::mir::StatementKind::PlaceMention(place.stable(tables)),
-                span,
-            },
-            AscribeUserType(place_projection, variance) => Statement {
-                kind: stable_mir::mir::StatementKind::AscribeUserType {
+                )
+            }
+            mir::StatementKind::SetDiscriminant { place, variant_index } => {
+                stable_mir::mir::StatementKind::SetDiscriminant {
+                    place: place.as_ref().stable(tables),
+                    variant_index: variant_index.stable(tables),
+                }
+            }
+            mir::StatementKind::Deinit(place) => {
+                stable_mir::mir::StatementKind::Deinit(place.stable(tables))
+            }
+
+            mir::StatementKind::StorageLive(place) => {
+                stable_mir::mir::StatementKind::StorageLive(place.stable(tables))
+            }
+
+            mir::StatementKind::StorageDead(place) => {
+                stable_mir::mir::StatementKind::StorageDead(place.stable(tables))
+            }
+            mir::StatementKind::Retag(retag, place) => {
+                stable_mir::mir::StatementKind::Retag(retag.stable(tables), place.stable(tables))
+            }
+            mir::StatementKind::PlaceMention(place) => {
+                stable_mir::mir::StatementKind::PlaceMention(place.stable(tables))
+            }
+            mir::StatementKind::AscribeUserType(place_projection, variance) => {
+                stable_mir::mir::StatementKind::AscribeUserType {
                     place: place_projection.as_ref().0.stable(tables),
                     projections: place_projection.as_ref().1.stable(tables),
                     variance: variance.stable(tables),
-                },
-                span,
-            },
-            Coverage(coverage) => {
-                Statement { kind: stable_mir::mir::StatementKind::Coverage(opaque(coverage)), span }
+                }
             }
-            Intrinsic(intrinstic) => Statement {
-                kind: stable_mir::mir::StatementKind::Intrinsic(intrinstic.stable(tables)),
-                span,
-            },
-            ConstEvalCounter => {
-                Statement { kind: stable_mir::mir::StatementKind::ConstEvalCounter, span }
+            mir::StatementKind::Coverage(coverage) => {
+                stable_mir::mir::StatementKind::Coverage(opaque(coverage))
             }
-            Nop => Statement { kind: stable_mir::mir::StatementKind::Nop, span },
+            mir::StatementKind::Intrinsic(intrinstic) => {
+                stable_mir::mir::StatementKind::Intrinsic(intrinstic.stable(tables))
+            }
+            mir::StatementKind::ConstEvalCounter => {
+                stable_mir::mir::StatementKind::ConstEvalCounter
+            }
+            mir::StatementKind::Nop => stable_mir::mir::StatementKind::Nop,
         }
     }
 }
@@ -844,76 +836,84 @@ impl<'tcx> Stable<'tcx> for mir::InlineAsmOperand<'tcx> {
 impl<'tcx> Stable<'tcx> for mir::Terminator<'tcx> {
     type T = stable_mir::mir::Terminator;
     fn stable(&self, tables: &mut Tables<'tcx>) -> Self::T {
-        use rustc_middle::mir::TerminatorKind::*;
         use stable_mir::mir::Terminator;
+        Terminator { kind: self.kind.stable(tables), span: self.source_info.span.stable(tables) }
+    }
+}
+
+impl<'tcx> Stable<'tcx> for mir::TerminatorKind<'tcx> {
+    type T = stable_mir::mir::TerminatorKind;
+    fn stable(&self, tables: &mut Tables<'tcx>) -> Self::T {
         use stable_mir::mir::TerminatorKind;
-        let span = self.source_info.span.stable(tables);
-        match &self.kind {
-            Goto { target } => {
-                Terminator { kind: TerminatorKind::Goto { target: target.as_usize() }, span }
+        match self {
+            mir::TerminatorKind::Goto { target } => {
+                TerminatorKind::Goto { target: target.as_usize() }
             }
-            SwitchInt { discr, targets } => Terminator {
-                kind: TerminatorKind::SwitchInt {
-                    discr: discr.stable(tables),
-                    targets: targets
-                        .iter()
-                        .map(|(value, target)| stable_mir::mir::SwitchTarget {
-                            value,
-                            target: target.as_usize(),
-                        })
-                        .collect(),
-                    otherwise: targets.otherwise().as_usize(),
-                },
-                span,
+            mir::TerminatorKind::SwitchInt { discr, targets } => TerminatorKind::SwitchInt {
+                discr: discr.stable(tables),
+                targets: targets
+                    .iter()
+                    .map(|(value, target)| stable_mir::mir::SwitchTarget {
+                        value,
+                        target: target.as_usize(),
+                    })
+                    .collect(),
+                otherwise: targets.otherwise().as_usize(),
             },
-            UnwindResume => Terminator { kind: TerminatorKind::Resume, span },
-            UnwindTerminate(_) => Terminator { kind: TerminatorKind::Abort, span },
-            Return => Terminator { kind: TerminatorKind::Return, span },
-            Unreachable => Terminator { kind: TerminatorKind::Unreachable, span },
-            Drop { place, target, unwind, replace: _ } => Terminator {
-                kind: TerminatorKind::Drop {
+            mir::TerminatorKind::UnwindResume => TerminatorKind::Resume,
+            mir::TerminatorKind::UnwindTerminate(_) => TerminatorKind::Abort,
+            mir::TerminatorKind::Return => TerminatorKind::Return,
+            mir::TerminatorKind::Unreachable => TerminatorKind::Unreachable,
+            mir::TerminatorKind::Drop { place, target, unwind, replace: _ } => {
+                TerminatorKind::Drop {
                     place: place.stable(tables),
                     target: target.as_usize(),
                     unwind: unwind.stable(tables),
-                },
-                span,
-            },
-            Call { func, args, destination, target, unwind, call_source: _, fn_span: _ } => {
-                Terminator {
-                    kind: TerminatorKind::Call {
-                        func: func.stable(tables),
-                        args: args.iter().map(|arg| arg.stable(tables)).collect(),
-                        destination: destination.stable(tables),
-                        target: target.map(|t| t.as_usize()),
-                        unwind: unwind.stable(tables),
-                    },
-                    span,
                 }
             }
-            Assert { cond, expected, msg, target, unwind } => Terminator {
-                kind: TerminatorKind::Assert {
+            mir::TerminatorKind::Call {
+                func,
+                args,
+                destination,
+                target,
+                unwind,
+                call_source: _,
+                fn_span: _,
+            } => TerminatorKind::Call {
+                func: func.stable(tables),
+                args: args.iter().map(|arg| arg.stable(tables)).collect(),
+                destination: destination.stable(tables),
+                target: target.map(|t| t.as_usize()),
+                unwind: unwind.stable(tables),
+            },
+            mir::TerminatorKind::Assert { cond, expected, msg, target, unwind } => {
+                TerminatorKind::Assert {
                     cond: cond.stable(tables),
                     expected: *expected,
                     msg: msg.stable(tables),
                     target: target.as_usize(),
                     unwind: unwind.stable(tables),
-                },
-                span,
-            },
-            InlineAsm { template, operands, options, line_spans, destination, unwind } => {
-                Terminator {
-                    kind: TerminatorKind::InlineAsm {
-                        template: format!("{template:?}"),
-                        operands: operands.iter().map(|operand| operand.stable(tables)).collect(),
-                        options: format!("{options:?}"),
-                        line_spans: format!("{line_spans:?}"),
-                        destination: destination.map(|d| d.as_usize()),
-                        unwind: unwind.stable(tables),
-                    },
-                    span,
                 }
             }
-            Yield { .. } | GeneratorDrop | FalseEdge { .. } | FalseUnwind { .. } => unreachable!(),
+            mir::TerminatorKind::InlineAsm {
+                template,
+                operands,
+                options,
+                line_spans,
+                destination,
+                unwind,
+            } => TerminatorKind::InlineAsm {
+                template: format!("{template:?}"),
+                operands: operands.iter().map(|operand| operand.stable(tables)).collect(),
+                options: format!("{options:?}"),
+                line_spans: format!("{line_spans:?}"),
+                destination: destination.map(|d| d.as_usize()),
+                unwind: unwind.stable(tables),
+            },
+            mir::TerminatorKind::Yield { .. }
+            | mir::TerminatorKind::GeneratorDrop
+            | mir::TerminatorKind::FalseEdge { .. }
+            | mir::TerminatorKind::FalseUnwind { .. } => unreachable!(),
         }
     }
 }

--- a/compiler/stable_mir/src/mir/body.rs
+++ b/compiler/stable_mir/src/mir/body.rs
@@ -191,7 +191,13 @@ pub enum NonDivergingIntrinsic {
 }
 
 #[derive(Clone, Debug)]
-pub enum Statement {
+pub struct Statement {
+    pub kind: StatementKind,
+    pub span: Span,
+}
+
+#[derive(Clone, Debug)]
+pub enum StatementKind {
     Assign(Place, Rvalue),
     FakeRead(FakeReadCause, Place),
     SetDiscriminant { place: Place, variant_index: VariantIdx },

--- a/tests/ui-fulldeps/stable-mir/crate-info.rs
+++ b/tests/ui-fulldeps/stable-mir/crate-info.rs
@@ -47,12 +47,12 @@ fn test_stable_mir(_tcx: TyCtxt<'_>) -> ControlFlow<()> {
     assert_eq!(body.blocks.len(), 1);
     let block = &body.blocks[0];
     assert_eq!(block.statements.len(), 1);
-    match &block.statements[0] {
-        stable_mir::mir::Statement::Assign(..) => {}
+    match &block.statements[0].kind {
+        stable_mir::mir::StatementKind::Assign(..) => {}
         other => panic!("{other:?}"),
     }
-    match &block.terminator {
-        stable_mir::mir::Terminator::Return => {}
+    match &block.terminator.kind {
+        stable_mir::mir::TerminatorKind::Return => {}
         other => panic!("{other:?}"),
     }
 
@@ -61,8 +61,8 @@ fn test_stable_mir(_tcx: TyCtxt<'_>) -> ControlFlow<()> {
     assert_eq!(body.locals.len(), 7);
     assert_eq!(body.blocks.len(), 4);
     let block = &body.blocks[0];
-    match &block.terminator {
-        stable_mir::mir::Terminator::Call { .. } => {}
+    match &block.terminator.kind {
+        stable_mir::mir::TerminatorKind::Call { .. } => {}
         other => panic!("{other:?}"),
     }
 
@@ -70,27 +70,27 @@ fn test_stable_mir(_tcx: TyCtxt<'_>) -> ControlFlow<()> {
     let body = types.body();
     assert_eq!(body.locals.len(), 6);
     assert_matches!(
-        body.locals[0].kind(),
+        body.locals[0].ty.kind(),
         stable_mir::ty::TyKind::RigidTy(stable_mir::ty::RigidTy::Bool)
     );
     assert_matches!(
-        body.locals[1].kind(),
+        body.locals[1].ty.kind(),
         stable_mir::ty::TyKind::RigidTy(stable_mir::ty::RigidTy::Bool)
     );
     assert_matches!(
-        body.locals[2].kind(),
+        body.locals[2].ty.kind(),
         stable_mir::ty::TyKind::RigidTy(stable_mir::ty::RigidTy::Char)
     );
     assert_matches!(
-        body.locals[3].kind(),
+        body.locals[3].ty.kind(),
         stable_mir::ty::TyKind::RigidTy(stable_mir::ty::RigidTy::Int(stable_mir::ty::IntTy::I32))
     );
     assert_matches!(
-        body.locals[4].kind(),
+        body.locals[4].ty.kind(),
         stable_mir::ty::TyKind::RigidTy(stable_mir::ty::RigidTy::Uint(stable_mir::ty::UintTy::U64))
     );
     assert_matches!(
-        body.locals[5].kind(),
+        body.locals[5].ty.kind(),
         stable_mir::ty::TyKind::RigidTy(stable_mir::ty::RigidTy::Float(
             stable_mir::ty::FloatTy::F64
         ))
@@ -100,8 +100,8 @@ fn test_stable_mir(_tcx: TyCtxt<'_>) -> ControlFlow<()> {
     let body = drop.body();
     assert_eq!(body.blocks.len(), 2);
     let block = &body.blocks[0];
-    match &block.terminator {
-        stable_mir::mir::Terminator::Drop { .. } => {}
+    match &block.terminator.kind {
+        stable_mir::mir::TerminatorKind::Drop { .. } => {}
         other => panic!("{other:?}"),
     }
 
@@ -109,15 +109,15 @@ fn test_stable_mir(_tcx: TyCtxt<'_>) -> ControlFlow<()> {
     let body = assert.body();
     assert_eq!(body.blocks.len(), 2);
     let block = &body.blocks[0];
-    match &block.terminator {
-        stable_mir::mir::Terminator::Assert { .. } => {}
+    match &block.terminator.kind {
+        stable_mir::mir::TerminatorKind::Assert { .. } => {}
         other => panic!("{other:?}"),
     }
 
     let monomorphic = get_item(&items, (DefKind::Fn, "monomorphic")).unwrap();
     for block in monomorphic.body().blocks {
-        match &block.terminator {
-            stable_mir::mir::Terminator::Call { func, .. } => match func {
+        match &block.terminator.kind {
+            stable_mir::mir::TerminatorKind::Call { func, .. } => match func {
                 stable_mir::mir::Operand::Constant(c) => match &c.literal.literal {
                     stable_mir::ty::ConstantKind::Allocated(alloc) => {
                         assert!(alloc.bytes.is_empty());
@@ -127,7 +127,7 @@ fn test_stable_mir(_tcx: TyCtxt<'_>) -> ControlFlow<()> {
                                 mut args,
                             )) => {
                                 let func = def.body();
-                                match func.locals[1]
+                                match func.locals[1].ty
                                     .fold(&mut args)
                                     .continue_value()
                                     .unwrap()
@@ -149,7 +149,7 @@ fn test_stable_mir(_tcx: TyCtxt<'_>) -> ControlFlow<()> {
                 },
                 other => panic!("{other:?}"),
             },
-            stable_mir::mir::Terminator::Return => {}
+            stable_mir::mir::TerminatorKind::Return => {}
             other => panic!("{other:?}"),
         }
     }

